File: android/src/main/java/cn/nodemedia/react_native_nodemediaclient/RCTNodePlayerView.java
//
//  RCTNodePlayerView.java
//
//  Created by Mingliang Chen on 2017/11/29.
//  Copyright © 2017-2024年 NodeMedia. All rights reserved.
//

package cn.nodemedia.react_native_nodemediaclient;

import android.widget.FrameLayout;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.events.RCTEventEmitter;

import cn.nodemedia.NodePlayer;

public class RCTNodePlayerView extends FrameLayout implements LifecycleEventListener {
    private NodePlayer np;
    private String url = "";

    public RCTNodePlayerView(ThemedReactContext context) {
        super(context);
        context.addLifecycleEventListener(this);
        np = new NodePlayer(context, RCTNodeMediaClient.getLicense());
        np.attachView(this);
        np.setOnNodePlayerEventListener((player, code, msg) -> {
            WritableMap event = Arguments.createMap();
            event.putInt("code", code);
            event.putString("msg", msg);
            ReactContext reactContext = (ReactContext) getContext();
            reactContext.getJSModule(RCTEventEmitter.class).receiveEvent(
                    getId(),
                    "topChange",
                    event);
        });
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public void setBufferTime(int bufferTime) {
        np.setBufferTime(bufferTime);
    }

    public void setScaleMode(int mode) {
        np.setScaleMode(mode);
    }

    public void setCryptoKey(String key) {
        np.setCryptoKey(key);
    }

    public void setHWAccelEnable(boolean HWAccelEnable) {
        np.setHWAccelEnable(HWAccelEnable);
    }

    public void setHTTPReferer(String HTTPReferer) {
        np.setHTTPReferer(HTTPReferer);
    }

    public void setHTTPUserAgent(String HTTPUserAgent) {
        np.setHTTPUserAgent(HTTPUserAgent);
    }

    public int start() {
        return np.start(this.url);
    }

    public int stop() {
        return np.stop();
    }

    @Override
    public void onHostResume() {

    }

    @Override
    public void onHostPause() {

    }

    @Override
    public void onHostDestroy() {
        np.stop();
        np.detachView();
    }
}

File: ios/RCTNodePlayerView.m
//
//  RCTNodePlayerView.m
//
//
//  Created by Mingliang Chen on 2017/11/29.
//  Copyright © 2017-2024年 NodeMedia. All rights reserved.
//

#import "RCTNodePlayerView.h"
#import "RCTNodeMediaClient.h"
#import <NodeMediaClient/NodeMediaClient.h>

@interface RCTNodePlayerView () <NodePlayerDelegate>

@property(strong, nonatomic) NodePlayer *np;

@end

@implementation RCTNodePlayerView

- (id)init {
  self = [super init];
  if (self) {
      _np = [[NodePlayer alloc] initWithLicense:[RCTNodeMediaClient license]];
      [_np setNodePlayerDelegate:self];
      [_np attachView:self];
  }
  return self;
}

- (void)dealloc
{
    [_np detachView];
}

-(void) onEventCallback:(nonnull id)sender event:(int)event msg:(nonnull NSString*)msg {
    if (_onChange) {
        _onChange(@{@"code": [NSNumber numberWithInteger:event], @"msg": msg});;
    }
}

- (void)setBufferTime:(int)bufferTime {
    [_np setBufferTime:bufferTime];
}

- (void)setVolume:(float)volume {
    [_np setVolume:volume];
}

- (void)setCryptoKey:(NSString *)cryptoKey {
    [_np setCryptoKey:cryptoKey];
}

- (void)setHWAccelEnable:(BOOL)HWAccelEnable {
    [_np setHWAccelEnable:HWAccelEnable];
}

- (void)setHTTPReferer:(NSString *)HTTPReferer {
    [_np setHTTPReferer:HTTPReferer];
}

- (void)setHTTPUserAgent:(NSString *)HTTPUserAgent {
    [_np setHTTPUserAgent:HTTPUserAgent];
}

- (void)setScaleMode:(int)scaleMode {
    [_np setScaleMode:scaleMode];
}

-(int)start{
    if(!_url) {
        return -1;
    }
    return [_np start:_url];
}

-(int)stop {
    return [_np stop];
}

@end
